test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java

```diff
@@ -0,0 +1,291 @@
+                + a + a + a + a + a + a + a + a + a + a
+                + a + a + a + a + a + a + a + a + a + a
+                + a + a; // a*42
+    }
+
+    @Test
+    @IR(failOn = IRNode.ADD_I)
+    @IR(counts = { IRNode.MUL_I, "1" })
+    private static int mulAndAddTo42(int a) {
+        return a * 40 + a + a; // a*41 + a => a*42
+    }
+
+    private static final int INT_MAX_MINUS_ONE = Integer.MAX_VALUE - 1;
+
+    @Test
+    @IR(failOn = IRNode.ADD_I)
+    @IR(counts = { IRNode.LSHIFT_I, "1", IRNode.SUB_I, "1" })
+    private static int mulAndAddToMax(int a) {
+        return a * INT_MAX_MINUS_ONE + a; // a*MAX => a*(MIN-1) => a*MIN - a => (a<<31) - a
+    }
+
+    @Test
+    @IR(failOn = IRNode.ADD_I)
+    @IR(counts = { IRNode.LSHIFT_I, "1" })
+    private static int mulAndAddToOverflow(int a) {
+        return a * Integer.MAX_VALUE + a; // a*(MAX+1) => a*(MIN) => a<<31
+    }
+
+    @Test
+    @IR(failOn = IRNode.ADD_I)
+    @IR(counts = { IRNode.CON_I, "1" })
+    private static int mulAndAddToZero(int a) {
+        return a * -1 + a; // 0
+    }
+
+    @Test
+    @IR(failOn = IRNode.ADD_I)
+    @IR(counts = { IRNode.LSHIFT_I, "1", IRNode.SUB_I, "1" })
+    private static int mulAndAddToMinus1(int a) {
+        return a * -2 + a; // a*-1 => a - (a<<1)
+    }
+
+    @Test
+    @IR(failOn = IRNode.ADD_I)
+    @IR(counts = { IRNode.MUL_I, "1" })
+    private static int mulAndAddToMinus42(int a) {
+        return a * -43 + a; // a*-42
+    }
+
+    // --- long tests ---
+    @Test
+    @IR(failOn = IRNode.ADD_L)
+    @IR(counts = { IRNode.LSHIFT_L, "1" })
+    private static long mulAndAddToIntOverflowL(long a) {
+        return a * Integer.MAX_VALUE + a; // a*(INT_MAX+1)
+    }
+
+    private static final long LONG_MAX_MINUS_ONE = Long.MAX_VALUE - 1;
+
+    @Test
+    @IR(failOn = IRNode.ADD_L)
+    @IR(counts = { IRNode.LSHIFT_L, "1", IRNode.SUB_L, "1" })
+    private static long mulAndAddToMaxL(long a) {
+        return a * LONG_MAX_MINUS_ONE + a; // a*MAX => a*(MIN-1) => a*MIN - 1 => (a<<63) - 1
+    }
+
+    @Test
+    @IR(failOn = IRNode.ADD_L)
+    @IR(counts = { IRNode.LSHIFT_L, "1" })
+    private static long mulAndAddToOverflowL(long a) {
+        return a * Long.MAX_VALUE + a; // a*(MAX+1) => a*(MIN) => a<<63
+    }
+
+    // --- bit shift tests ---
+    @Test
+    @IR(failOn = {IRNode.ADD_I, IRNode.LSHIFT_I})
+    private static int bitShiftToOverflow() {
+        int i, x = 0;
+        for (i = 0; i < 32; i++) {
+            x = i;
+        }
+
+        // x = 31 (phi), i = 32 (phi + 1)
+        return i + (x << i) + i; // Expects 32 + 31 + 32 = 95
+    }
+
+    @Test
+    @IR(failOn = {IRNode.ADD_L, IRNode.LSHIFT_L})
+    private static long bitShiftToOverflowL() {
+        int i, x = 0;
+        for (i = 0; i < 64; i++) {
+            x = i;
+        }
+
+        // x = 63 (phi), i = 64 (phi + 1)
+        return i + (x << i) + i; // Expects 64 + 63 + 64 = 191
+    }
+}
```

Comment:
Would it make sense to add some randomized patterns, just for result verification?

You can use `Generators.java` to get interesting values. Of course that would mean not doing IR verification, but at least it would give us better confidence that the values are correct.

I'm imagining expressions like this:
`return a * CON1 + a * CON2 + a * CON3 + a * CON4`

Where the CON are defined as a `public static final` field with a random value generated by `Generators`.

The advantage of using `Generators` is that it generates powers-of-two more frequently, which seems to be relevant here.

Reply:
That's a very valid suggestion. Thanks. I updated the test. Please take a look if I understood your intention correctly.

However, it does look like this `a * CON1 + a * CON2 + ...` pattern is picked up by existing the associative optimization first:

https://github.com/openjdk/jdk/blob/a23fb0af65f491ef655ba114fcc8032a09a55213/src/hotspot/share/opto/addnode.cpp#L345-L377